src/test/java/io/aiven/kafka/auth/AivenSaslScramServerCallbackHandlerTest.java

```diff
@@ -0,0 +1,61 @@
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+import io.aiven.kafka.auth.AivenSaslScramServerCallbackHandler;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.security.auth.login.AppConfigurationEntry;
+import org.apache.kafka.common.security.scram.ScramCredential;
+import org.apache.kafka.common.security.scram.ScramLoginModule;
+import org.junit.Test;
+
+public class AivenSaslScramServerCallbackHandlerTest {
+  static final String USERS_JSON = "[{\"username\":\"testuser\",\"password\":\"testpassword\"}]";
+
+  @Test
+  public void testAivenSaslPlainServerCallbackHandler() throws IOException {
+    Path tempPath = Files.createTempDirectory("test-aiven-kafka-sasl-scram-handler");
+    Path configFilePath = Paths.get(tempPath.toString(), "sasl_passwd.json");
+
+    File passwdJson = new File(configFilePath.toString());
+
+    Files.write(configFilePath, USERS_JSON.getBytes());
+
+    Map<String, String> entryConfigs = new HashMap<String, String>();
+    entryConfigs.put("users.config", configFilePath.toString());
+    AppConfigurationEntry entry = new AppConfigurationEntry(ScramLoginModule.class.getName(),
+            AppConfigurationEntry.LoginModuleControlFlag.REQUIRED, entryConfigs);
+    List<AppConfigurationEntry> jaasConfigs = new ArrayList<AppConfigurationEntry>();
+    jaasConfigs.add(entry);
+
+    // scram-sha-256
+    AivenSaslScramServerCallbackHandler handler = new AivenSaslScramServerCallbackHandler();
+    handler.configure(null, "SCRAM-SHA-256", jaasConfigs);
+
+    ScramCredential creds = handler.getScramCreds("testuser");
```

Comment:
I'd prefer testing the whole public method instead of some implementation details. It's not super-convenient here, because it's callbacks. But it can be done with mocking, like:

```java
package io.aiven.kafka.auth;

import static org.mockito.ArgumentMatchers.argThat;
import static org.mockito.ArgumentMatchers.isNull;
import static org.mockito.Mockito.times;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import javax.security.auth.callback.Callback;
import javax.security.auth.callback.NameCallback;
import javax.security.auth.callback.UnsupportedCallbackException;
import javax.security.auth.login.AppConfigurationEntry;

import org.apache.kafka.common.security.scram.ScramCredentialCallback;
import org.apache.kafka.common.security.scram.ScramLoginModule;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.MockitoJUnitRunner;

@RunWith(MockitoJUnitRunner.class)
public class AivenSaslScramServerCallbackHandlerTest {
  static final String USERS_JSON = "[{\"username\":\"testuser\",\"password\":\"testpassword\"}]";

  @Mock
  NameCallback nameCallback;
  @Mock
  ScramCredentialCallback scramCredentialCallback;

  Callback[] callbacks;

  static List<AppConfigurationEntry> jaasConfigs;

  /**
   * Set up class.
   */
  @BeforeClass
  public static void setUpClass() throws IOException {
    Path tempPath = Files.createTempDirectory("test-aiven-kafka-sasl-scram-handler");
    Path configFilePath = Paths.get(tempPath.toString(), "sasl_passwd.json");

    Files.write(configFilePath, USERS_JSON.getBytes());

    Map<String, String> entryConfigs = new HashMap<String, String>();
    entryConfigs.put("users.config", configFilePath.toString());
    AppConfigurationEntry entry = new AppConfigurationEntry(ScramLoginModule.class.getName(),
        AppConfigurationEntry.LoginModuleControlFlag.REQUIRED, entryConfigs);
    jaasConfigs = new ArrayList<>();
    jaasConfigs.add(entry);
  }

  @Before
  public void setUp() {
    callbacks = new Callback[]{ nameCallback, scramCredentialCallback };
  }

  @Test
  public void testScramSha256() throws IOException, UnsupportedCallbackException {
    test("SCRAM-SHA-256");
  }

  @Test
  public void testScramSha512() throws IOException, UnsupportedCallbackException {
    test("SCRAM-SHA-512");
  }

  private void test(String mechanism) throws IOException, UnsupportedCallbackException {
    AivenSaslScramServerCallbackHandler handler = new AivenSaslScramServerCallbackHandler();
    handler.configure(null, mechanism, jaasConfigs);

    Mockito.doReturn("testuser").when(nameCallback).getDefaultName();
    handler.handle(callbacks);
    Mockito.verify(scramCredentialCallback, times(1))
        .scramCredential(argThat(cred -> cred.iterations() == 4096));

    Mockito.clearInvocations(callbacks);

    Mockito.doReturn("invaliduser").when(nameCallback).getDefaultName();
    handler.handle(callbacks);
    Mockito.verify(scramCredentialCallback, times(1))
        .scramCredential(isNull());
  }
}
```

Reply:
This looks nice, thanks. But perhaps it would be worthwhile to look at the testing of all aspects of the module in a separate PR?

Reply:
Yeah, sure